pkg/gather/service/analyze.go

```diff
@@ -118,7 +118,6 @@ func checkAPIURLs(a analysis) bool {
 	}
 	// Note: Even when there is a stage failure, we are not returning false here. That is
 	// intentional because we donot want to report this as an error in the "analyze" output.
-	logrus.Warn("The bootstrap machine is unable to resolve API and/or API-Int Server URLs")
```

Comment:
I still don't see anything in this function that's specific to the API resolution steps, but the function name is still `checkAPIURLs`.  If the goal is to report the last error in the `bootkube` steps, regardless of whether that is an API-resolution error or not, can we rename the function to `checkBootkubeService` or something, and adjust the function's Godocs to match?   Also I'm unclear about why `checkReleaseImageDownload` returns `false` on error while this function returns `true`.  Perhaps there  should be a generic `checkService` function?  Or maybe I'm missing the nuance of  the `Note: Even when there is a stage failure...` comment?

Reply:
1. None of the functions in https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootstrap-verify-api-server-urls.sh actually return an error even when the URL is detected as invalid or it is not resolvable. Hence, that comment.
But, that also means that we will not be able to detect the right time to display the logrus.Warn(). +1 for removing it.

2. I think we can also remove the "record-stage-*" messages from https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootstrap-verify-api-server-urls.sh. The msgs generated from that are not really adding any value. 

3. +1 to renaming the function to checkBootkubeService() with corresponding changes to the Godocs. That way, there is no assumptions about the checks happening in the bootkube service.

Reply:
> +1 to have api-dns-check.done so that we run these checks only once. (In the custom-dns case, these checks may not succeed right away. So, we skip this test in that case, or allow multiple checks.)

The check would be run on each iteration until it passes, once it passes it would be skipped thereafter. Agreed that it would not work for the `api` record in custom DNS because it is expected that the record is created post install.

Reply:
Since analyze just looks at the logs generated during bootkube service, this should still work for custom DNS because both API and API-Int entries are created within the CoreDNS pod. 
I am not aware of all the reasons why DNS resolution can fluctuate during install (causing this erroneous reporting we are seeing here), but I suspect the same reasons would be in play even when custom dns is enabled.

Reply:
> should still work for custom DNS because both API and API-Int entries are created within the CoreDNS pod.

Ah, that’s right. So we should be good. 👍

> I am not aware of all the reasons why DNS resolution can fluctuate during install 

it’s not dns resolution that’s failing—the dns check curls the api server which becomes unavailable periodically during an install.